### PR TITLE
Issue 172: optimize map hasKey() and hasEntry()

### PR DESCRIPTION
**Pull request by brownian-motion**
Implements the enhancement in hamcrest/JavaHamcrest#172

Instead of a linear search over all entries, this change leverages Map.containsKey(K key) to check that a map hasKey() or hasEntry().

### CODE REVIEW

File: hamcrest/src/main/java/org/hamcrest/collection/IsMapContaining.java [hunk 1]
```diff
@@ -72,9 +72,32 @@ public void describeTo(Description description) {
...
+        public IsMapContainingEntry(K key, V value)
+        {
+            super(equalTo(key), equalTo(value));
+            this.key = key;
+        }
+
+        @Override
+        public boolean matchesSafely(Map<? extends K, ? extends V> map)
+        {
+            return map.containsKey(key) && super.valueMatcher.matches(map.get(key));
```
Comment (nibsi) [inline hamcrest/src/main/java/org/hamcrest/collection/IsMapContaining.java:97]: There are maps that don't allow null keys and (stupidly) throw an exception when you try to check for its existence. You need to account for them here.
```suggestion
            try {
              return map.containsKey(key) && super.valueMatcher.matches(map.get(key));
            catch (NullPointerException ex) {
              return false;
            }
```

**Comment by brownian-motion** [comment]
Thank you for your feedback, @nibsi . I'm afraid I created the branch from my local `master` branch, which included a different change, instead of `origin/master`; I'm embarrassed I missed that before pushing. I'll rebase and push a PR without the unrelated change.

**Comment by brownian-motion** [comment]
I've added some tests that cover the relevant edge cases covered by the code addition. I believe this commit is ready for final review + merging if approved.

**Comment by sf105** [comment]
Interesting question, the null thing is not something I thought of. The question is what is the right behaviour when using null with an inappropriate collection. Personally, I would have expected it to fail, rather than mismatch, as that's what would happen in the production code. I fear that we'll be hiding failure modes. 

Also, is that a copy'n'paste in the comment?

**Comment by brownian-motion** [comment]
@sf105 There's one case I can think of where the safe API is better: asserting that a null key is absent:

`assertThat(new Hashtable<Object, Object>(), not(hasKey(null)));`

I would expect this to pass and continue to the next assertion; allowing a null pointer exception would force you to avoid touching particular types/code paths or start to put conditional logic and try/catches in your tests.



Also, sorry, where do you mean? I'm not sure what copy-and-paste you mean.

**Comment by brownian-motion** [comment]
@sf105 I would be happy to make whichever further changes you think would benefit the project the best. When convenient, please let me know what further adjustments might be appropriate here.

**Comment by nhojpatrick** [comment]
@brownian-motion looks good, please can you rebase from master, as `hamcrest-core` and `hamcrest-library` have been refactored a lot and also deprecated, so that everything is just in `hamcrest`.

**Comment by brownian-motion** [comment]
Hello again! I know it's been two years, but I've rebased this PR and would be happy to help with any further steps to get it ready for merging.

@nhojpatrick please feel free to look through this again when convenient :)

**Comment by nhojpatrick** [comment]
Going to try and kick start hamcrest, so if you want to get it merged, please rebase from the branch `v2.3-candidates`.
Still trying to understand how has permissions to perform a release.
